### PR TITLE
[libc++][hardening] Constrain construction for `__{(static_)bounded,wrap}_iter`

### LINKED ISSUE #50058: libc++ permits conversions between std::vector<>::iterators of distinct types
|  |  |
| --- | --- |
| Bugzilla Link | [50714](https://llvm.org/bz50714) |
| Version | 12.0 |
| OS | Linux |
| Reporter | LLVM Bugzilla Contributor |
| CC | @dwblaikie,@mclow |

## Extended Description 
Hello!

Recently I've noticed that, libc++ permits conversions
between std::vector<>::iterators of two different element_types,
as long as pointers to these element_types are convertible:

    struct foo       { float f; };
    struct bar : foo { int   i; };

    auto vec = std::vector
    {
        bar { foo { 1.1f }, 7 },
        bar { foo { 0.8f }, 5 },  
    };

    typename std::vector<foo>::iterator it = vec.begin();
    //                   ^
    // note: The value_type of the iterator 'it' is foo, not bar!

    ++it;

    auto fp_value = it->f; // bug: iterator 'it' is not valid!

This is very dangerous, because the iterator 'it' expects
to iterate over a continuous block of memory, containing only foo objects:

    +-----+-----+-----
    | foo | foo | foo ...
    +-----+-----+-----
    ^     ^     ^
    it    it+1  it+2  ...

However, this is not what is actually happening in memory:

    +-----------+-----------+-------
    |    bar    |    bar    |    bar ...
    +-----------+-----------+-------
    | foo |     | foo |     | foo |  ...
    +-----------+-----------+-------
    ^     ^     ^
    it    it+1  it+2                 ...

This contradiction makes std::next(it) is not dereferenceable
and will result in undefined behavior (UB),
when an object will be accessed though the iterator 'it'.

Compiler Explorer link: https://godbolt.org/z/5zzb7dTPd

Correct me if I'm wrong,
but ideally this bug should be reported at compile-time by a C++ compiler.

BTW, this is what other C++ standard library implementations are already doing:
 - [GCC] libstdc++ -> https://godbolt.org/z/bqsfW4aMb
 - [MSVC] STL      -> https://godbolt.org/z/4TnK45jcM

Also, the ISO C++ standard only requires that std::vector<>::iterator
must be convertible to std::vector<>::const_iterator,
so other conversions could be disabled, right?

    Table 73: Container requirements [tab:container.req]
    +-------------+---------------+------------------------------------+
    |  Expression |  Return type  |                 Note               |
    +-------------+---------------+------------------------------------+
    |             | iterator type | any iterator category that meets   |
    | X::iterator | whose value   | the forward iterator requirements. |
    |             | type is T     | convertible to X::const_iterator   |
    +-------------+---------------+------------------------------------+

ISO C++20 Working Draft:
 ~ http://open-std.org/jtc1/sc22/wg21/docs/papers/2020/n4861.pdf

Why this conversion is even allowed by libc++?

From what I've investigated, in the libc++ implementation,
the std::vector<>::iterator is an alias to the __wrap_iter<> class template:

    template <class _Tp, class _Allocator /* = allocator<_Tp> */>
    class _LIBCPP_TEMPLATE_VIS vector
    {
    ...
        typedef __wrap_iter<pointer>                     iterator;
        typedef __wrap_iter<const_pointer>               const_iterator;
    ...
    };

[GitHub] llvm-project/libcxx/include/vector:
 ~ https://github.com/llvm/llvm-project/blob/main/libcxx/include/vector#L488

Looking at the __wrap_iter<> class template revels the reason why
this dangerous conversion was even allowed by the Clang compiler:

    template <class _Iter>
    class __wrap_iter
    {
    ...
        template <class _Up> ...
        __wrap_iter(const __wrap_iter<_Up>& __u,
                    std::enable_if_t<
                    std::is_convertible_v<_Up, iterator_type>>* = nullptr)
        ... { ... }
    ...
    };

[GitHub] llvm-project/libcxx/include/iterator:
 ~ https://github.com/llvm/llvm-project/blob/main/libcxx/include/iterator#L1437

Essentially, as long as type_x* is convertible to type_y*,
std::vector<type_x>::iterator is convertible to std::vector<type_y>::iterator.

Every time I see such a weakly-constrained, non-explicit conversion constructor,
red flags are immediately being raised in my brain, since
this could lead to quite surprising results and can potentially cause issues ...

Do you think it would be beneficial to further constrain
conversions of std::vector<>::iterator?

Would it make sense to allow only conversion
from std::vector<>::iterator to std::vector<>::const_iterator?

Thank you, Mateusz Zych

### PR DESCRIPTION
**Pull request by frederick-vs-ja**
This PR restricts construction to cases where reference types of source/destination iterators are (`T&`, `T&`) or (`T&`, `const T&`) ( where `T` can be const).

Fixes #50058.

### CODE REVIEW

**Comment by frederick-vs-ja** [comment]
Timed out in `std/utilities/charconv/charconv.msvc/test.pass.cpp`, which seems unrelated.

https://github.com/llvm/llvm-project/actions/runs/11718999636/job/32643141533

**Comment by philnik777** [comment]
Please make these separate PRs. The changes are unrelated AFAICT.

**Comment by frederick-vs-ja** [comment]
> Please make these separate PRs. The changes are unrelated AFAICT.

Separated to #115304.

File: libcxx/include/__iterator/bounded_iter.h [hunk 1]
```diff
@@ -67,7 +72,13 @@ struct __bounded_iter {
   _LIBCPP_HIDE_FROM_ABI __bounded_iter(__bounded_iter const&) = default;
   _LIBCPP_HIDE_FROM_ABI __bounded_iter(__bounded_iter&&)      = default;
 
-  template <class _OtherIterator, __enable_if_t< is_convertible<_OtherIterator, _Iterator>::value, int> = 0>
+  template < class _OtherIterator,
+             __enable_if_t<
+                 _And<is_constructible<_Iterator, const _OtherIterator&>,
+                      is_convertible<const _OtherIterator&, _Iterator>,
```
Comment (ldionne) [inline libcxx/include/__iterator/bounded_iter.h:78]: Do you need to check both `is_convertible` and `is_constructible`? Doesn't convertibility subsume constructibility?

File: libcxx/include/__iterator/bounded_iter.h [hunk 2]
```diff
@@ -67,7 +72,13 @@ struct __bounded_iter {
...
+                      _Or<is_same<reference, __iter_reference<_OtherIterator> >,
+                          is_same<reference, __make_const_lvalue_ref<__iter_reference<_OtherIterator> > > > >::value,
```
[the remark this answers is not in the document]
Comment (frederick-vs-ja) [reply libcxx/include/__iterator/bounded_iter.h:80]: Yes. [[iterator.concept.contiguous]](https://eel.is/c++draft/iterator.concept.contiguous) and [[allocator.requirements.general]/92](https://eel.is/c++draft/allocator.requirements.general#92) (for possibly fancy pointers, note that `std::addressof` rejects rvalues) require this.

File: libcxx/include/__iterator/bounded_iter.h (hunk 1 above)
Comment (frederick-vs-ja) [reply libcxx/include/__iterator/bounded_iter.h:78]: I'll change back to use `is_convertible` only...

However, [allocator.requirements.general] (and perhaps many other places in the library specification) sometimes only says "(implicit) conversion"/"(implicitly) convertible", while C++ types can be so tricky that explicit and implicit conversions can have different semantics, or the explicit conversion can be ill-formed while the implicit one is not. Maybe an LWG issue, requiring conversions between fancy pointers to be sane, is desired.

**Comment by frederick-vs-ja** [comment]
> Can you also make this change for `__static_bounded_iter` which was added just now?

Changed. Also added a previously missed change to one friend declaration.

**Review by ldionne — APPROVED** [review]

**Pull request merged**